cmd/kubeadm/app/cmd/join.go

```diff
@@ -308,44 +309,55 @@ func newJoinOptions() *joinOptions {
 	// Re-apply defaults to the public kubeadm API (this will set only values not exposed/not set as a flags)
-	kubeadmscheme.Scheme.Default(options.externalcfg)
+	kubeadmscheme.Scheme.Default(opt.externalcfg)
 
 	// Validate standalone flags values and/or combination of flags and then assigns
 	// validated values to the public kubeadm config API when applicable
 
 	// if a token is provided, use this value for both discovery-token and tls-bootstrap-token when those values are not provided
-	if len(options.token) > 0 {
-		if len(options.externalcfg.Discovery.TLSBootstrapToken) == 0 {
-			options.externalcfg.Discovery.TLSBootstrapToken = options.token
+	if len(opt.token) > 0 {
+		if len(opt.externalcfg.Discovery.TLSBootstrapToken) == 0 {
+			opt.externalcfg.Discovery.TLSBootstrapToken = opt.token
 		}
-		if len(options.externalcfg.Discovery.BootstrapToken.Token) == 0 {
-			options.externalcfg.Discovery.BootstrapToken.Token = options.token
+		if len(opt.externalcfg.Discovery.BootstrapToken.Token) == 0 {
+			opt.externalcfg.Discovery.BootstrapToken.Token = opt.token
 		}
 	}
 
 	// if a file or URL from which to load cluster information was not provided, unset the Discovery.File object
-	if len(options.externalcfg.Discovery.File.KubeConfigPath) == 0 {
-		options.externalcfg.Discovery.File = nil
+	if len(opt.externalcfg.Discovery.File.KubeConfigPath) == 0 {
+		opt.externalcfg.Discovery.File = nil
 	}
 
 	// if an APIServerEndpoint from which to retrive cluster information was not provided, unset the Discovery.BootstrapToken object
 	if len(args) == 0 {
-		options.externalcfg.Discovery.BootstrapToken = nil
+		opt.externalcfg.Discovery.BootstrapToken = nil
 	} else {
-		if len(options.cfgPath) == 0 && len(args) > 1 {
+		if len(opt.cfgPath) == 0 && len(args) > 1 {
 			klog.Warningf("[join] WARNING: More than one API server endpoint supplied on command line %v. Using the first one.", args)
 		}
-		options.externalcfg.Discovery.BootstrapToken.APIServerEndpoint = args[0]
+		opt.externalcfg.Discovery.BootstrapToken.APIServerEndpoint = args[0]
 	}
 
 	// if not joining a control plane, unset the ControlPlane object
-	if !options.controlPlane {
-		options.externalcfg.ControlPlane = nil
+	if !opt.controlPlane {
+		opt.externalcfg.ControlPlane = nil
 	}
 
-	ignorePreflightErrorsSet, err := validation.ValidateIgnorePreflightErrors(options.ignorePreflightErrors)
+	// if the admin.conf file already exists, use it for skipping the discovery process.
+	// NB. this case can happen when we are joining a control-plane node  only (and phases are invoked atomically)
+	var tlsBootstrapCfg *clientcmdapi.Config
+	if _, err := os.Stat(kubeadmconstants.GetAdminKubeConfigPath()); err == nil && opt.controlPlane {
+		// use the admin.conf as tlsBootstrapCfg, that is the kubeconfig file used for reading the kubeadm-config during discovery
+		tlsBootstrapCfg, err = clientcmd.LoadFromFile(kubeadmconstants.GetAdminKubeConfigPath())
```

Comment:
does this print more details in `klog.V(x > 0)` mode?
do we have a sufficiently descriptive message here if the command fails (e.g. empty admin.confg file).

Reply:
Good point
Added logs + custom error management.